File: macos/CursorPocketMac/Tests/CursorPocketMacKitTests/GoldenContractTests.swift
import Foundation
import XCTest
@testable import CursorPocketMacKit

/// Reads the cross-platform golden fixtures in `spec/capture-manifest/`.
/// The Windows test suite reads the exact same files, so either side
/// drifting from the shared storage contract breaks a CI.
final class GoldenContractTests: XCTestCase {
    private func specDirectory() throws -> URL {
        // Walk up from this source file until the repo root (the directory
        // containing `spec`) — the tests run from an arbitrary build folder.
        var url = URL(fileURLWithPath: #filePath)
        while url.pathComponents.count > 1 {
            url.deleteLastPathComponent()
            let spec = url.appendingPathComponent("spec", isDirectory: true)
            var isDirectory: ObjCBool = false
            if FileManager.default.fileExists(atPath: spec.path, isDirectory: &isDirectory),
               isDirectory.boolValue {
                return spec.appendingPathComponent("capture-manifest", isDirectory: true)
            }
        }
        // A skip here would let contract drift pass silently, so fail hard.
        struct SpecDirectoryNotFound: Error {}
        XCTFail("Repo root with spec/ not found above \(#filePath)")
        throw SpecDirectoryNotFound()
    }

    private func goldenLines() throws -> [String] {
        let url = try specDirectory().appendingPathComponent("golden.jsonl")
        let text = try String(contentsOf: url, encoding: .utf8)
        return text.split(whereSeparator: \.isNewline).map(String.init)
    }

    func testGoldenManifestCoversEveryKindAndDecodesWithTheRealRecord() throws {
        let lines = try goldenLines()
        XCTAssertEqual(lines.count, 5)
        let decoder = JSONDecoder()
        var records: [CaptureRecord] = []
        for line in lines {
            records.append(try decoder.decode(CaptureRecord.self, from: Data(line.utf8)))
        }
        XCTAssertEqual(
            records.map(\.captureKind),
            [.screenshot, .video, .audio, .text, .link])
        for record in records {
            XCTAssertEqual(record.schemaVersion, 2)
            XCTAssertFalse(record.id.isEmpty)
            XCTAssertFalse(record.preview.isEmpty)
            // Windows-shaped values: forward-slash relative paths under a
            // dated folder, in the kind's category, with the kind's extension.
            XCTAssertFalse(record.relativePath.contains("\\"))
            XCTAssertTrue(record.relativePath.hasPrefix("2026-08-18/"))
            XCTAssertTrue(record.relativePath.contains("/\(record.captureKind.category)/"))
            XCTAssertTrue(record.relativePath.hasSuffix(record.captureKind.fileExtension))
            XCTAssertTrue(CaptureStore.isSafeRelativePath(record.relativePath))
            // .NET "O" timestamps (seven fractional digits, numeric offset)
            // must parse to a real instant.
            XCTAssertNotEqual(record.created, .distantPast, "createdAt failed to parse: \(record.createdAt)")
        }
    }

    func testGoldenMetadataValueShapesSurvive() throws {
        let decoder = JSONDecoder()
        let records = try goldenLines().map {
            try decoder.decode(CaptureRecord.self, from: Data($0.utf8))
        }
        XCTAssertEqual(records[0].metadata["width"], .number(1920))
        XCTAssertEqual(records[0].metadata["height"], .number(1080))
        XCTAssertEqual(records[0].metadata["recovered"], .bool(false))
        XCTAssertEqual(records[1].metadata["duration_seconds"], .number(42.5))
        XCTAssertEqual(records[1].metadata["recovered"], .bool(true))
        XCTAssertEqual(records[2].metadata["duration_seconds"], .number(14.2))
        XCTAssertTrue(records[3].metadata.isEmpty)
        XCTAssertEqual(records[4].metadata["url"], .string("https://www.example.com/docs/page"))
        XCTAssertEqual(records[4].metadata["host"], .string("example.com"))
    }

    func testReencodingKeepsTheExactWireKeySet() throws {
        let decoder = JSONDecoder()
        let encoder = JSONEncoder()
        for line in try goldenLines() {
            let original = try XCTUnwrap(
                try JSONSerialization.jsonObject(with: Data(line.utf8)) as? [String: Any])
            let record = try decoder.decode(CaptureRecord.self, from: Data(line.utf8))
            let reencoded = try XCTUnwrap(
                try JSONSerialization.jsonObject(with: encoder.encode(record)) as? [String: Any])
            XCTAssertEqual(Set(reencoded.keys), Set(original.keys), "Key drift in line: \(line)")
            let originalMetadata = try XCTUnwrap(original["metadata"] as? [String: Any])
            let reencodedMetadata = try XCTUnwrap(reencoded["metadata"] as? [String: Any])
            XCTAssertEqual(Set(reencodedMetadata.keys), Set(originalMetadata.keys))
        }
    }

    func testGoldenUrlBodyParsesWithLinkCapture() throws {
        let url = try specDirectory().appendingPathComponent("golden.url")
        let body = try String(contentsOf: url, encoding: .utf8)
        XCTAssertTrue(body.hasPrefix("[InternetShortcut]"))
        XCTAssertEqual(
            LinkCapture.url(fromInternetShortcut: body),
            "https://www.example.com/docs/page")
        // The canonical body is exactly what this app would write back.
        XCTAssertEqual(
            LinkCapture.internetShortcutBody(for: "https://www.example.com/docs/page"), body)
    }
}

File: native/CursorPocket.Tests/GoldenContractTests.cs
using System.Text.Json;
using CursorPocket.Core.Models;

namespace CursorPocket.Tests;

// The macOS test suite (GoldenContractTests.swift) reads the exact same
// files in spec/capture-manifest/, so either platform drifting from the
// shared storage contract breaks a CI.
public sealed class GoldenContractTests
{
    private static string FixturePath(string name) =>
        Path.Combine(AppContext.BaseDirectory, "Fixtures", name);

    private static string[] GoldenLines() =>
        File.ReadAllLines(FixturePath("golden.jsonl"))
            .Where(line => line.Length > 0)
            .ToArray();

    private static CaptureRecord Deserialize(string line) =>
        JsonSerializer.Deserialize<CaptureRecord>(line)
            ?? throw new InvalidOperationException($"Golden line deserialized to null: {line}");

    [Fact]
    public void GoldenManifestCoversEveryKindAndDeserializesWithTheRealRecord()
    {
        var lines = GoldenLines();
        Assert.Equal(5, lines.Length);
        var records = lines.Select(Deserialize).ToArray();

        Assert.Equal(
            [CaptureKind.Screenshot, CaptureKind.Video, CaptureKind.Audio, CaptureKind.Text, CaptureKind.Link],
            records.Select(record => record.CaptureKind));

        foreach (var record in records)
        {
            Assert.Equal(2, record.SchemaVersion);
            Assert.False(string.IsNullOrEmpty(record.Id));
            Assert.False(string.IsNullOrEmpty(record.Preview));
            // Forward-slash relative paths under a dated folder — never
            // backslashes, so both platforms resolve the same file.
            Assert.DoesNotContain('\\', record.RelativePath);
            Assert.StartsWith("2026-08-18/", record.RelativePath, StringComparison.Ordinal);
            // .NET "O" timestamps must parse to a real instant.
            Assert.NotEqual(DateTimeOffset.MinValue, record.Created);
        }
    }

    [Fact]
    public void GoldenMetadataValueShapesSurvive()
    {
        var records = GoldenLines().Select(Deserialize).ToArray();

        Assert.Equal(1920, records[0].Metadata["width"].GetInt32());
        Assert.Equal(1080, records[0].Metadata["height"].GetInt32());
        Assert.False(records[0].Metadata["recovered"].GetBoolean());
        Assert.Equal(42.5, records[1].Metadata["duration_seconds"].GetDouble(), 3);
        Assert.True(records[1].Metadata["recovered"].GetBoolean());
        Assert.Equal(14.2, records[2].Metadata["duration_seconds"].GetDouble(), 3);
        Assert.Empty(records[3].Metadata);
        Assert.Equal("https://www.example.com/docs/page", records[4].Metadata["url"].GetString());
        Assert.Equal("example.com", records[4].Metadata["host"].GetString());
    }

    [Fact]
    public void ReserializingKeepsTheExactWireKeySet()
    {
        foreach (var line in GoldenLines())
        {
            var record = Deserialize(line);
            using var original = JsonDocument.Parse(line);
            using var reserialized = JsonDocument.Parse(JsonSerializer.Serialize(record));

            // Sorted so the assertion is order-independent and certain.
            static string[] Keys(JsonElement element) =>
                element.EnumerateObject()
                    .Select(property => property.Name)
                    .OrderBy(name => name, StringComparer.Ordinal)
                    .ToArray();

            Assert.Equal(Keys(original.RootElement), Keys(reserialized.RootElement));
            Assert.Equal(
                Keys(original.RootElement.GetProperty("metadata")),
                Keys(reserialized.RootElement.GetProperty("metadata")));
        }
    }

    [Fact]
    public void GoldenUrlBodyIsTheCanonicalInternetShortcut()
    {
        var body = File.ReadAllText(FixturePath("golden.url"));
        Assert.StartsWith("[InternetShortcut]", body, StringComparison.Ordinal);
        Assert.Contains("URL=https://www.example.com/docs/page", body, StringComparison.Ordinal);

        var linkRecord = Deserialize(GoldenLines()[4]);
        Assert.Equal(linkRecord.Metadata["url"].GetString(), ReadShortcutUrl(body));
    }

    // Mirrors the parse both apps apply to a `.url` body.
    private static string? ReadShortcutUrl(string body)
    {
        foreach (var line in body.Split('\n', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            if (line.StartsWith("URL=", StringComparison.OrdinalIgnoreCase))
            {
                var value = line[4..].Trim();
                return value.Length == 0 ? null : value;
            }
        }
        return null;
    }
}
